File: http-server/src/main/java/org/zunpeng/vertx/service/mysql/MysqlServiceImpl.java
package org.zunpeng.vertx.service.mysql;

import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonObject;
import io.vertx.mutiny.mysqlclient.MySQLClient;
import io.vertx.mutiny.sqlclient.Pool;
import io.vertx.mutiny.sqlclient.templates.RowMapper;
import io.vertx.mutiny.sqlclient.templates.SqlTemplate;
import io.vertx.sqlclient.Row;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.zunpeng.vertx.service.mysql.domain.Device;

import java.util.*;

public class MysqlServiceImpl implements MysqlService {

  private static final Logger logger = LogManager.getLogger(MysqlServiceImpl.class);

  private final Pool pool;

  public MysqlServiceImpl(Pool pool, Handler<AsyncResult<MysqlService>> handler) {
    this.pool = pool;
    SqlTemplate.forQuery(pool, "select 1")
      .mapTo(new RowMapper<JsonObject>(Row::toJson))
      .execute(new HashMap<>())
      .subscribe()
      .with(rowSet -> {
        logger.info("------- result of test: {}", rowSet.iterator().next());
        handler.handle(Future.succeededFuture(this));
      }, throwable -> {
        logger.error(throwable.getMessage(), throwable);
        handler.handle(Future.failedFuture(throwable));
      });


  }

  @Override
  public MysqlService getDeviceById(Long id, Handler<AsyncResult<Device>> handler) {
    String sql = "select * from device where id = #{id}";
    Map<String, Object> parameterMap = Collections.singletonMap("id", id);
    SqlTemplate.forQuery(pool, sql)
      .mapTo(Device.class)
      .execute(parameterMap)
      .subscribe()
      .with(rowSet -> {
        if (rowSet.size() == 0) {
          handler.handle(Future.succeededFuture(null));
          return;
        }
        if (rowSet.size() > 1) {
          handler.handle(Future.failedFuture(new RuntimeException("size of result is error")));
          return;
        }
        handler.handle(Future.succeededFuture(rowSet.iterator().next()));
      }, throwable -> {
        logger.error(throwable.getMessage(), throwable);
        handler.handle(Future.failedFuture(throwable));
      });
    return this;
  }

  @Override
  public MysqlService listDevice(Handler<AsyncResult<List<Device>>> handler) {
    String sql = "select * from device";
    SqlTemplate.forQuery(pool, sql)
      .mapTo(Device.class)
      .execute(new HashMap<>())
      .subscribe()
      .with(rowSet -> {
        List<Device> deviceList = new ArrayList<>();
        for (Device device : rowSet) {
          deviceList.add(device);
        }
        handler.handle(Future.succeededFuture(deviceList));
      }, throwable -> {
        logger.error(throwable.getMessage(), throwable);
        handler.handle(Future.failedFuture(throwable));
      });
    return this;
  }

  @Override
  public MysqlService saveDevice(String sn, Handler<AsyncResult<Device>> handler) {
    String sql = "insert into device(sn) values (#{sn})";
    SqlTemplate.forUpdate(pool, sql)
      .execute(Collections.singletonMap("sn", sn))
      .subscribe()
      .with(rowSet -> {
        long lastInsertId = rowSet.property(MySQLClient.LAST_INSERTED_ID);
        logger.info("lastInsertId: {}, {} row(s) updated!", lastInsertId, rowSet.rowCount());
        handler.handle(Future.succeededFuture(new Device(lastInsertId, sn)));
      }, throwable -> {
        logger.error(throwable.getMessage(), throwable);
        handler.handle(Future.failedFuture(throwable));
      });
    return this;
  }

  @Override
  public MysqlService updateDevice(Long id, String sn, Handler<AsyncResult<Void>> handler) {
    String sql = "update device set sn = #{sn} where id = #{id}";
    Map<String, Object> paramMap = new HashMap<>();
    paramMap.put("id", id);
    paramMap.put("sn", sn);
    SqlTemplate.forUpdate(pool, sql)
      .execute(paramMap)
      .subscribe()
      .with(rowSet -> {
        logger.info("{} row(s) updated!", rowSet.rowCount());
        handler.handle(Future.succeededFuture());
      }, throwable -> {
        logger.error(throwable.getMessage(), throwable);
        handler.handle(Future.failedFuture(throwable));
      });
    return this;
  }

  @Override
  public MysqlService txUpdateDevice(Long id, String sn, Handler<AsyncResult<Void>> handler) {
    String sql = "update device set sn = #{sn} where id = #{id}";
    Map<String, Object> paramMap = new HashMap<>();
    paramMap.put("id", id);
    paramMap.put("sn", sn);
    pool.withTransaction(conn ->
      SqlTemplate.forUpdate(conn, sql)
        .execute(paramMap)
        // 模拟失败，触发回滚
        // .onItem()
        // .failWith(a -> {
        //   logger.info("{} row(s) updated!", a.rowCount());
        //   return new RuntimeException("error");
        // })
    )
      .subscribe()
      .with(a -> {
        logger.info("{} row(s) updated!", a.rowCount());
        handler.handle(Future.succeededFuture());
      }, t -> {
        logger.error(t.getMessage(), t);
        handler.handle(Future.failedFuture(t));
      });
    return this;
  }
}

File: service-proxy-demo/src/main/kotlin/com/duobeiyun/feishu/service/mysql/MysqlServiceImpl.kt
package com.duobeiyun.feishu.service.mysql

import io.vertx.core.AsyncResult
import io.vertx.core.Future
import io.vertx.core.Handler
import io.vertx.core.Vertx
import io.vertx.core.json.JsonObject
import io.vertx.sqlclient.Pool
import io.vertx.sqlclient.Row
import io.vertx.sqlclient.templates.SqlTemplate
import mu.KotlinLogging

class MysqlServiceImpl(
  private val vertx: Vertx,
  private val poolFromCom: Pool,
  private val poolFromNet: Pool,
  handle: Handler<AsyncResult<MysqlService>>
) : MysqlService {

  companion object {
    private val logger = KotlinLogging.logger {}
  }

  init {
    handle.handle(Future.succeededFuture(this))
  }

  override fun countByConvertStatusFromCom(query: JsonObject, handle: Handler<AsyncResult<JsonObject>>): MysqlService {
    val sql = """
      select convertStatus, count(*) as a_count from nebula_meeting.course_info
      where endTime between #{startTime} and #{endTime} group by convertStatus order by convertStatus asc
    """.trimIndent()
    countByConvertStatus(sql = sql, queryMap = query.map, pool = poolFromCom, handle = handle)
    return this
  }

  override fun countByConvertStatusFromNet(query: JsonObject, handle: Handler<AsyncResult<JsonObject>>): MysqlService {
    val sql = """
      select convert_status as convertStatus, count(*) as a_count from sail_platform_course.course
      where end_time between #{startTime} and #{endTime} group by convert_status order by convert_status asc
    """.trimIndent()
    countByConvertStatus(sql = sql, queryMap = query.map, pool = poolFromNet, handle = handle)
    return this
  }

  fun countByConvertStatus(
    sql: String,
    queryMap: Map<String, Any>,
    pool: Pool,
    handle: Handler<AsyncResult<JsonObject>>
  ) {
    SqlTemplate.forQuery(pool, sql)
      .mapTo(Row::toJson)
      .execute(queryMap)
      .onSuccess { rowSet ->
        if (rowSet == null || rowSet.size() == 0) {
          logger.info { "query: ${JsonObject(queryMap).encode()}, course not found" }
          handle.handle(Future.succeededFuture(JsonObject()))
        } else {
          val result = rowSet.associate {
            val convertStatus = when (it.getInteger("convertStatus", 0)) {
              -1 -> "default"
              0 -> "processing"
              1 -> "success"
              2 -> "failure"
              3 -> "empty"
              else -> {
                logger.error { "convert status not found" }
                "error"
              }
            }
            convertStatus to it.getLong("a_count", 0)
          }
          logger.info { "result: $result" }
          handle.handle(Future.succeededFuture(JsonObject(result)))
        }
      }.onFailure { throwable ->
        logger.error(throwable) { throwable.message }
        handle.handle(Future.failedFuture(throwable))
      }
  }
}
